### PR TITLE
Apply OpenShift OAuth provider

### LINKED ISSUE #15670: Make it possible to get OpenShift oauth token from theia IDE container
### Is your task related to a problem? Please describe.

We need an OpenShift authentication token to inject it to Openshift connector plugin if Che is started on OpenShift.

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
Add OpenShift OAuth provider to be able to retrieve openshift token.

### What issues does this PR fix or reference?
fixes #15670

### CODE REVIEW

File: wsmaster/che-core-api-auth-openshift/src/main/java/org/eclipse/che/security/oauth/OpenShiftModule.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import com.google.inject.AbstractModule;
+import com.google.inject.multibindings.Multibinder;
+import org.eclipse.che.inject.DynaModule;
+
+/**
+ * Setup OpenShiftOAuthAuthenticator in guice container.
+ *
+ * @author Igor Vinokur
+ */
+@DynaModule
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-auth-openshift/src/main/java/org/eclipse/che/security/oauth/OpenShiftModule.java]: Can you explicitly install this module in ws-master assembly?

Comment (vinokurig) [reply wsmaster/che-core-api-auth-openshift/src/main/java/org/eclipse/che/security/oauth/OpenShiftModule.java]: done

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 2]
```diff
@@ -161,6 +161,11 @@ che.oauth.github.authuri= https://github.com/login/oauth/authorize
 che.oauth.github.tokenuri= https://github.com/login/oauth/access_token
 che.oauth.github.redirecturis= http://localhost:${CHE_PORT}/api/oauth/callback
 
+# You can setup GitHub OAuth to automate authentication to remote repositories.
```
Comment (skabashnyuk) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Can you update content? It looks like a copy-paste from Github.

**Comment by skabashnyuk** [comment]
CC @tolusha 
Do we need any changes in che-operator?

File: wsmaster/che-core-api-auth-openshift/src/main/java/org/eclipse/che/security/oauth/OpenShiftModule.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.google.inject.AbstractModule;
+import com.google.inject.multibindings.Multibinder;
+
+/**
+ * Setup OpenShiftOAuthAuthenticator in guice container.
+ *
+ * @author Igor Vinokur
+ */
+public class OpenShiftModule extends AbstractModule {
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-auth-openshift/src/main/java/org/eclipse/che/security/oauth/OpenShiftModule.java]: Can you rename it to OpenOAuthShiftModule?

File: wsmaster/che-core-api-auth-openshift/src/main/java/org/eclipse/che/security/oauth/OpenShiftOAuthAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  public final String getOAuthProvider() {
+    return "openshift";
+  }
+
+  @Override
+  public OAuthToken getToken(String userId) throws IOException {
+    final OAuthToken token = super.getToken(userId);
+    // Check if the token is valid for requests.
+    if (!(token == null || token.getToken() == null || token.getToken().isEmpty())) {
+      String tokenVerifyUrl = this.openshiftEndpoint + "oapi/v1/projects";
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-auth-openshift/src/main/java/org/eclipse/che/security/oauth/OpenShiftOAuthAuthenticator.java]: have you checked this with OpenShift 3.9(not sure we need it), 3.11, 4,2 ?

**Comment by tolusha** [comment]
@skabashnyuk 
I hope @vinokurig  will test changes with operator to figure that out.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 5]
```diff
@@ -162,7 +162,7 @@ che.oauth.github.tokenuri= https://github.com/login/oauth/access_token
 che.oauth.github.redirecturis= http://localhost:${CHE_PORT}/api/oauth/callback
 
 # You can setup GitHub OAuth to automate authentication to remote repositories.
-# You need to first register this application with GitHub OAuth.
+# You need to first register this application with OpenShift OAuth.
```
Comment (skabashnyuk) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: >You need to first register this application with OpenShift OAuth.

Are you sure this is a good idea?

Comment (vinokurig) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Removed the leftovers

Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: I believe we need some explanation of what is it and how to get it.

Comment (vinokurig) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Updated the description

**Comment by skabashnyuk** [comment]
Can you also provide a docs pr with the explanation of how this feature should be used in practice? What are the prerequisites?
-  How it should work with/without multiuser mode. 
- How it should work with/without OpenShift OAuth mode enabled

**Comment by tolusha** [comment]
@skabashnyuk 
> Do we need any changes in che-operator?

I don't think so.
The introduced properties can be added on `CustomCheProperties` of the CR without changing che-operator
Do you have any concerns ?

**Review by mshaposhnik — APPROVED** [review]

**Comment by vinokurig** [comment]
@davidfestal 
>Do we need any changes in che-operator?
>>
>>I don't think so.
The introduced properties can be added on CustomCheProperties of the CR without changing che-operator
Do you have any concerns ?

Can we use `CustomCheProperties` or do we need to explicitly define the properties in the che-operator?

File: wsmaster/che-core-api-auth-openshift/src/main/java/org/eclipse/che/security/oauth/OpenShiftOAuthAuthenticator.java (hunk 4 above)
Comment (vinokurig) [reply wsmaster/che-core-api-auth-openshift/src/main/java/org/eclipse/che/security/oauth/OpenShiftOAuthAuthenticator.java]: Added a property `che.oauth.openshift.verify_token_url` to be able to configure the request url for  different OpenShift versions

**Comment by vinokurig** [comment]
@skabashnyuk 
>Can you also provide a docs pr with the explanation of how this feature should be used in practice? What are the prerequisites?
>
>How it should work with/without multiuser mode.

For that I need to be sure that https://github.com/eclipse/che/pull/15705 will be merged soon. As for now it works only on single user mode
> How it should work with/without OpenShift OAuth mode enabled

I don't see any problems with `OpenShift OAuth mode`. The che OpenShift OAuth provider requires creating an own OpenShift OAuth client https://docs.openshift.com/container-platform/4.3/authentication/configuring-internal-oauth.html#oauth-register-additional-client_configuring-internal-oauth

**Comment by vinokurig** [comment]
@skabashnyuk 
> Can you also provide a docs pr with the explanation of how this feature should be used in practice? What are the prerequisites?
>
>How it should work with/without multiuser mode.
How it should work with/without OpenShift OAuth mode enabled

Added a docs PR: https://github.com/eclipse/che-docs/pull/1076

**Review by skabashnyuk — APPROVED** [review]

**Review by tolusha — APPROVED** [review]

**Comment by vinokurig** [comment]
crw-ci-test

**Comment by vinokurig** [comment]
[ci-test]

**Comment by davidfestal** [comment]
> I don't see any problems with OpenShift OAuth mode. The che OpenShift OAuth provider requires creating an own OpenShift OAuth client https://docs.openshift.com/container-platform/4.3/authentication/configuring-internal-oauth.html#oauth-register-additional-client_configuring-internal-oauth

Not sure I followed the whole story about this issue, but does this mean that now, by default, users will have to manually create an additional OAuth client in Openshift in addition to the one already automatically created by the Che operator when Openshift OAuth is enabled (to support Keycloak-Openshift identity provider) ?

**Review by davidfestal — CHANGES_REQUESTED** [review]
Please  could you answer the question I asked here: https://github.com/eclipse/che/pull/15963#issuecomment-591500848 ?

Thanks !

**Comment by vinokurig** [comment]
@davidfestal 
We have similar situation with the GitHub oAuth provider,  it requires creating the GitHub oAuth client on the GitHub side. We can automate creating the OpenShift oAuth client by the operator but it will mean that the OpenShift oAuth provider is supported only for operator installation.

**Comment by davidfestal** [comment]
@vinokurig 

> We have similar situation with the GitHub oAuth provider, it requires creating the GitHub oAuth client on the GitHub side.

GitHub is a bit different since, afaik, you cannot automate the creation of the GitHub OAuth client easily, and the GitHub OAuth client is somehow something expected to be owned by the Che admin user in a self-service mode.

OTOH setting up an OpenShift OAuth client requires cluster admin rights into the cluster, and the installer user of Che/CRW on OpenShift doesn't necessarily own the cluster on which it installs Che (think of OSD addon for example). So this should be done by the Che operator, and is already done to support Keycloak Openshift identity provider setup.

> We can automate creating the OpenShift oAuth client by the operator but it will mean that the OpenShift oAuth provider is supported only for operator installation.

Afaik the operator-based installation is **the official installation mode** for OpenShift, either through `chectl` for OpenShift 3.11, or though OLM + OperatorHub for OpenShift 4.x (+ OSD, indirectly via addon).
Of course, for anyone having cluster admin rights, and wanting to use raw yaml files to deploy Che on OpenShift, he could still create the OAuth Client by hand by following the doc.

But it seems that *requiring a **manual setup** of an additional OAuth client inside the OpenShift cluster before being able to use the OpenShift integration tools (OpenShift plugin ?) will be a **no-go** in a number of contexts we aim to support*.

**Review by l0rd — APPROVED** [review]
After discussing with @vinokurig I consider that the user experience doesn't get worse with this PR. If we merge this PR Che will still use OpenShift OAuth as identity provider even if the administrator doesn't create the extra OpenShift OAuth client. But the administrator would have a new (manual) option to enable the OpenShift OAuth provider for the workspaces. 

For this reason I consider it as a good first step and the PR can be merged as it is (and even cherry picked to 7.9.x if we have time cc @nickboldt). But as mentioned by @davidfestal there are some important improvements that need to be introduced (update the operator to automatically create the client for instance). Such improvements can be done in a second PR.

**Comment by vinokurig** [comment]
@l0rd @davidfestal Created sub-issue: https://github.com/eclipse/che/issues/16199

**Comment by vinokurig** [comment]
crw-ci-test

**Comment by vinokurig** [comment]
[ci-test]

**Comment by vinokurig** [comment]
crw-ci-test

**Comment by davidfestal** [comment]
> After discussing with @vinokurig I consider that the user experience doesn't get worse with this PR.

> For this reason I consider it as a good first step and the PR can be merged as it is 

Thanks @l0rd for providing more context.

> Created sub-issue: #16199

Thanks @vinokurig

Let me approve this PR then.

**Review by davidfestal — APPROVED** [review]

**Comment by vinokurig** [comment]
crw-ci-test

**Comment by vinokurig** [comment]
crw-ci-test

**Comment by dmytro-ndp** [comment]
crw-ci-test

**Comment by dmytro-ndp** [comment]
crw-ci-test

**Pull request merged**